File: system-test/dotnet/Dsl.Core/UseCase/External/Erp/UseCases/ReturnsPromotion.cs
using Driver.Port.External.Erp;
using Driver.Port.External.Erp.Dtos;
using Driver.Port.External.Erp.Dtos.Error;
using Dsl.Core.External.Erp.UseCases.Base;
using Common;
using Dsl.Core.Shared;

namespace Dsl.Core.External.Erp.UseCases;

public class ReturnsPromotion : BaseErpCommand<VoidValue, VoidVerification>
{
    private bool _promotionActive;
    private string? _discount;

    public ReturnsPromotion(IErpDriver driver, UseCaseContext context)
        : base(driver, context)
    {
    }

    public ReturnsPromotion WithActive(bool promotionActive)
    {
        _promotionActive = promotionActive;
        return this;
    }

    public ReturnsPromotion WithDiscount(string? discount)
    {
        _discount = discount;
        return this;
    }

    public ReturnsPromotion WithDiscount(decimal discount)
    {
        return WithDiscount(discount.ToString());
    }

    public override async Task<ErpUseCaseResult<VoidValue, VoidVerification>> Execute()
    {
        var request = new ReturnsPromotionRequest
        {
            PromotionActive = _promotionActive,
            Discount = _discount
        };

        var result = await _driver.ReturnsPromotionAsync(request);

        return new ErpUseCaseResult<VoidValue, VoidVerification>(
            result,
            _context,
            (response, ctx) => new VoidVerification(response, ctx));
    }
}

File: system-test/java/src/main/java/com/optivem/shop/testkit/core/usecase/external/erp/usecases/ReturnsPromotion.java
package com.optivem.shop.testkit.core.usecase.external.erp.usecases;

import com.optivem.shop.testkit.driver.port.external.erp.ErpDriver;
import com.optivem.shop.testkit.driver.port.external.erp.dtos.ReturnsPromotionRequest;
import com.optivem.shop.testkit.core.usecase.external.erp.usecases.base.BaseErpUseCase;
import com.optivem.shop.testkit.common.Converter;
import com.optivem.shop.testkit.core.shared.UseCaseResult;
import com.optivem.shop.testkit.core.shared.UseCaseContext;
import com.optivem.shop.testkit.core.shared.VoidVerification;

public class ReturnsPromotion extends BaseErpUseCase<Void, VoidVerification> {
    private boolean promotionActive;
    private String discount;

    public ReturnsPromotion(ErpDriver driver, UseCaseContext context) {
        super(driver, context);
    }

    public ReturnsPromotion withActive(boolean promotionActive) {
        this.promotionActive = promotionActive;
        return this;
    }

    public ReturnsPromotion withDiscount(String discount) {
        this.discount = discount;
        return this;
    }

    public ReturnsPromotion withDiscount(double discount) {
        return withDiscount(Converter.fromDouble(discount));
    }

    @Override
    public UseCaseResult<Void, VoidVerification> execute() {
        var request = ReturnsPromotionRequest.builder()
                .promotionActive(promotionActive)
                .discount(discount)
                .build();

        var result = driver.returnsPromotion(request);

        return new UseCaseResult<>(result, context, VoidVerification::new);
    }
}
